File: exercism/c/binary-search/binary_search.c
#include "binary_search.h"

const int *binary_search(int value, const int *arr, const size_t length) {

  if (!arr || length == 0 || value < arr[0] || value > arr[length - 1])
    return NULL;

  if (length == 1)
    return arr;

  size_t start = 0, end = length, middle;

  while (start < end) {
    middle = (start + end) / 2; // half: [start, end)
    if (arr[middle] == value)
      return (const int *)(arr + middle);
    else if (arr[middle] > value)
      end = middle;
    else
      start = middle + 1;
  }

  return NULL;
}

File: exercism/python/binary-search/binary_search.py
from typing import List


def find(search_list: List[int], value: int) -> int:
    if not isinstance(search_list, list):
        raise ValueError("search_list is not a list")

    start = 0
    end = len(search_list) - 1

    while start <= end:
        middle = (start + end) // 2
        item = search_list[middle]

        if item == value:
            return middle

        if value > item:
            start = middle + 1
        else:
            end = middle - 1

    raise ValueError("value not in array")
